File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/java/org/rgn/first/mapreduce/actors/MapActor.java
package org.rgn.first.mapreduce.actors;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

import org.rgn.first.mapreduce.messages.MapData;
import org.rgn.first.mapreduce.messages.WordCount;

import akka.actor.UntypedActor;

public class MapActor extends UntypedActor {
	String[] STOP_WORDS = { "a", "am", "an", "and", "are", "as", "at", "be",
			"do", "go", "if", "in", "is", "it", "of", "on", "the", "to" };
	List<String> STOP_WORDS_LIST = Arrays.asList(STOP_WORDS);

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof String) {
			String work = (String) message;
			getSender().tell(evaluateExpression(work), getSelf());
		} else {
			unhandled(message);
		}
	}

	private MapData evaluateExpression(String line) {
		List<WordCount> dataList = new ArrayList<>();

		StringTokenizer parser = new StringTokenizer(line);
		while (parser.hasMoreTokens()) {
			String word = parser.nextToken().toLowerCase();
			if (!STOP_WORDS_LIST.contains(word)) {
				dataList.add(new WordCount(word, 1));
			}
		}

		return new MapData(dataList);
	}

}

File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/scala/org/rgn/scala/first/mapreduce/actors/MapActor.scala
package org.rgn.scala.first.mapreduce.actors

import akka.actor.Actor
import org.rgn.scala.first.mapreduce.MapData
import scala.collection.mutable.ArrayBuffer
import org.rgn.scala.first.mapreduce.WordCount

class MapActor extends Actor {
  val STOP_WORDS_LIST = List("a", "am", "an", "and", "are", "as",
    "at", "be", "do", "go", "if", "in", "is", "it", "of", "on", "the",
    "to")

  def receive: Receive = {
    case message: String =>
      sender ! evaluateExpression(message)
  }

  def evaluateExpression(line: String): MapData = MapData {
    line.split("""\s+""").foldLeft(ArrayBuffer.empty[WordCount]) {
      (index, word) => if(!STOP_WORDS_LIST.contains(word.toLowerCase)) {
        index += WordCount(word.toLowerCase, 1)
      }else
        index
    }
  }
}